# Phase 9 Visual Implementation Guide

## ğŸ¯ Three Strategies at a Glance

```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                   FALLBACK STRATEGY ENFORCEMENT                 â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

STRATEGY 1: qdrant_first (DEFAULT - â­ RECOMMENDED)
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                                                                â”‚
â”‚   START                                                        â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ TRY: Qdrant Semantic Search ğŸ”                       â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ SUCCESS? ğŸ“Š                                          â”‚
â”‚     â”‚   â”œâ”€ YES â†’ Use Qdrant Context                          â”‚
â”‚     â”‚   â”‚        Extract with targeted results               â”‚
â”‚     â”‚   â”‚        âœ“ Complete successfully                     â”‚
â”‚     â”‚   â”‚                                                   â”‚
â”‚     â”‚   â””â”€ NO â†’ âš ï¸ Fallback to Raw Text                    â”‚
â”‚     â”‚          Extract with full text context              â”‚
â”‚     â”‚          âœ“ Complete successfully                    â”‚
â”‚     â”‚                                                     â”‚
â”‚     â””â”€â†’ RESULT: âœ… Always Succeeds                        â”‚
â”‚                                                           â”‚
â”‚   USE WHEN: Production deployments (best effort)         â”‚
â”‚   LATENCY: Variable (Qdrant + fallback time if needed)   â”‚
â”‚   QUALITY: High (semantic when possible)                 â”‚
â”‚                                                           â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

STRATEGY 2: qdrant_only (STRICT - â­ FOR VALIDATION)
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                                                                â”‚
â”‚   START                                                        â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ TRY: Qdrant Semantic Search ğŸ”                       â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ SUCCESS? ğŸ“Š                                          â”‚
â”‚     â”‚   â”œâ”€ YES â†’ Use Qdrant Context                          â”‚
â”‚     â”‚   â”‚        Extract with targeted results               â”‚
â”‚     â”‚   â”‚        âœ“ Complete successfully                     â”‚
â”‚     â”‚   â”‚                                                   â”‚
â”‚     â”‚   â””â”€ NO â†’ âŒ ABORT                                   â”‚
â”‚     â”‚          Fail with error message                     â”‚
â”‚     â”‚          âœ— Extraction aborted                       â”‚
â”‚     â”‚                                                     â”‚
â”‚     â””â”€â†’ RESULT: âœ… or âŒ (no middle ground)              â”‚
â”‚                                                           â”‚
â”‚   USE WHEN: Validation & testing (strict requirements)   â”‚
â”‚   LATENCY: Fast (but fails fast if no results)          â”‚
â”‚   QUALITY: High (semantic search only)                   â”‚
â”‚                                                           â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

STRATEGY 3: raw_only (BASELINE - â­ FOR DEBUGGING)
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                                                                â”‚
â”‚   START                                                        â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ SKIP: Qdrant Entirely âš™ï¸                             â”‚
â”‚     â”‚   (No Qdrant dependency)                              â”‚
â”‚     â”‚                                                         â”‚
â”‚     â”œâ”€â†’ USE: Raw Text Context                               â”‚
â”‚     â”‚   Load full text from pages_text dict                 â”‚
â”‚     â”‚                                                         â”‚
â”‚     â””â”€â†’ RESULT: âœ… Always Succeeds                         â”‚
â”‚                                                              â”‚
â”‚   USE WHEN: Debugging, comparison, rapid iteration       â”‚
â”‚   LATENCY: Fast (no Qdrant overhead)                       â”‚
â”‚   QUALITY: Medium (full text, less targeted)              â”‚
â”‚                                                            â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

---

## ğŸ”„ Execution Flow Comparison

```
SCENARIO 1: Qdrant is Available & Has Results
â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                     qdrant_first                         â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ğŸ” Search Qdrant                                        â”‚
â”‚ ğŸ¯ Found 3 results (similarity: 0.92, 0.88, 0.85)     â”‚
â”‚ ğŸ“Š Using Qdrant context for extraction                 â”‚
â”‚ âœ… Extraction succeeded with targeted data              â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                    qdrant_only                          â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ğŸ” Search Qdrant                                        â”‚
â”‚ ğŸ¯ Found 3 results (similarity: 0.92, 0.88, 0.85)     â”‚
â”‚ ğŸ“Š Using Qdrant context for extraction                 â”‚
â”‚ âœ… Extraction succeeded with targeted data              â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                     raw_only                            â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ âš™ï¸  Strategy 'raw_only' - skipping Qdrant search        â”‚
â”‚ ğŸ“ Loading raw text context                             â”‚
â”‚ âœ… Extraction succeeded with full text                  â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜


SCENARIO 2: Qdrant Unavailable or No Results
â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                     qdrant_first                         â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ğŸ” Search Qdrant                                        â”‚
â”‚ âŒ No results found (timeout or unavailable)           â”‚
â”‚ âš ï¸  Fallback: Using raw text instead                    â”‚
â”‚ âœ… Extraction succeeded with fallback                   â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                    qdrant_only                          â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ ğŸ” Search Qdrant                                        â”‚
â”‚ âŒ No results found (timeout or unavailable)           â”‚
â”‚ âŒ Strategy 'qdrant_only': ABORTING                     â”‚
â”‚ âœ— Extraction FAILED - no fallback allowed               â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                     raw_only                            â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚ âš™ï¸  Strategy 'raw_only' - skipping Qdrant search        â”‚
â”‚ ğŸ“ Loading raw text context                             â”‚
â”‚ âœ… Extraction succeeded with full text                  â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

---

## ğŸ“Š Decision Matrix

```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚  Choose Your Strategy Based on Your Needs                        â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚                                                                  â”‚
â”‚  Need Semantic Search Quality?                                  â”‚
â”‚  â”œâ”€ YES â†’ Use qdrant_first (default) or qdrant_only            â”‚
â”‚  â””â”€ NO  â†’ Use raw_only for pure LLM testing                    â”‚
â”‚                                                                  â”‚
â”‚  Can Tolerate Fallback?                                         â”‚
â”‚  â”œâ”€ YES â†’ Use qdrant_first (production)                        â”‚
â”‚  â”œâ”€ NO  â†’ Use qdrant_only (strict)                             â”‚
â”‚  â””â”€ SKIP QDRANT â†’ Use raw_only (baseline)                      â”‚
â”‚                                                                  â”‚
â”‚  Need to Validate Qdrant?                                      â”‚
â”‚  â”œâ”€ YES â†’ Use qdrant_only (will fail if not working)          â”‚
â”‚  â””â”€ NO  â†’ Use qdrant_first (auto-handles failures)            â”‚
â”‚                                                                  â”‚
â”‚  Production or Testing?                                         â”‚
â”‚  â”œâ”€ PRODUCTION â†’ Use qdrant_first (recommended)               â”‚
â”‚  â”œâ”€ VALIDATION â†’ Use qdrant_only (strict)                     â”‚
â”‚  â””â”€ DEBUGGING  â†’ Use raw_only (baseline)                      â”‚
â”‚                                                                  â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

---

## ğŸ” Code Implementation Pattern

```python
# All 6 extraction functions follow this pattern:

def extract_[TYPE](...) -> [Result]:
    """Extract [TYPE] with fallback strategy enforcement."""
    
    global FALLBACK_STRATEGY
    
    # 1. DETERMINE STRATEGY
    if FALLBACK_STRATEGY == 'raw_only':
        # Strategy: Skip Qdrant entirely
        logger.info(f"âš™ï¸  Strategy 'raw_only' - skipping Qdrant")
        context_text = raw_text_data
    
    else:
        # Strategy: Try Qdrant first
        
        # 2. ATTEMPT QDRANT SEARCH
        context_docs = []
        for query in search_queries:
            docs = search_qdrant_for_context(...)
            context_docs.extend(docs)
        
        # 3. EVALUATE RESULTS
        if context_docs:
            # Success case
            logger.info("ğŸ“Š Using Qdrant context")
            context_text = build_context(context_docs)
        
        elif FALLBACK_STRATEGY == 'qdrant_only':
            # Strict failure case
            logger.error("âŒ Strategy 'qdrant_only' - ABORTING")
            raise ValueError("No Qdrant context available")
        
        else:  # qdrant_first
            # Graceful fallback case
            logger.warning("âš ï¸  Fallback to raw text")
            context_text = raw_text_data
    
    # 4. CONTINUE EXTRACTION
    result = llm.extract(context_text)
    return result
```

---

## ğŸ“ˆ Performance & Quality Comparison

```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                      QDRANT AVAILABLE                          â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚                                                                â”‚
â”‚ Strategy         Speed      Quality   Qdrant   Fallback      â”‚
â”‚ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€    â”€â”€â”€â”€â”€â”€     â”€â”€â”€â”€â”€â”€â”€â”€  â”€â”€â”€â”€â”€â”€â”€  â”€â”€â”€â”€â”€â”€â”€â”€      â”‚
â”‚ qdrant_first     ğŸŸ¢ Fast    ğŸŸ¢ High   âœ“ Used   âœ— Unused     â”‚
â”‚ qdrant_only      ğŸŸ¢ Fast    ğŸŸ¢ High   âœ“ Used   âœ— Unused     â”‚
â”‚ raw_only         ğŸŸ¡ Slow    ğŸŸ  Med    âœ— Skip   âœ“ Always     â”‚
â”‚                                                                â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                   QDRANT UNAVAILABLE                           â”‚
â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤
â”‚                                                                â”‚
â”‚ Strategy         Speed      Quality   Qdrant   Fallback      â”‚
â”‚ â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€    â”€â”€â”€â”€â”€â”€     â”€â”€â”€â”€â”€â”€â”€â”€  â”€â”€â”€â”€â”€â”€â”€  â”€â”€â”€â”€â”€â”€â”€â”€      â”‚
â”‚ qdrant_first     ğŸŸ¡ Slow    ğŸŸ  Med    âœ— Fail   âœ“ Active     â”‚
â”‚ qdrant_only      ğŸ”´ FAIL    âœ— NONE    âœ— Fail   âœ— None      â”‚
â”‚ raw_only         ğŸŸ¢ Fast    ğŸŸ  Med    âœ— Skip   âœ“ Always     â”‚
â”‚                                                                â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜

Legend:
ğŸŸ¢ Good  |  ğŸŸ¡ OK  |  ğŸŸ  Medium  |  ğŸ”´ Bad/Fail
```

---

## ğŸ¯ Real-World Examples

### Example 1: Production Setup (qdrant_first)

```bash
$ python src/rag/structured_extraction.py \
    --company-slug world_labs

Fallback strategy: qdrant_first
...
ğŸ” Searching Qdrant: "company overview mission vision"
ğŸ¯ Found 3 results with scores: 0.923, 0.891, 0.876
ğŸ“Š Using Qdrant context for company extraction
âœ“ Successfully extracted company: World Labs
âœ“ Saved to: data/structured/world-labs.json
```

### Example 2: Validation Mode (qdrant_only)

```bash
$ python src/rag/structured_extraction.py \
    --company-slug world_labs \
    --fallback-strategy qdrant_only

Fallback strategy: qdrant_only
...
ğŸ” Searching Qdrant: "company overview mission vision"
âŒ Qdrant connection failed
âŒ Strategy 'qdrant_only': No Qdrant results - ABORTING

Error: No Qdrant context available and 'qdrant_only' strategy selected
â†’ This tells us Qdrant is not working!
```

### Example 3: Baseline Mode (raw_only)

```bash
$ python src/rag/structured_extraction.py \
    --company-slug world_labs \
    --fallback-strategy raw_only

Fallback strategy: raw_only
...
âš™ï¸  Strategy 'raw_only' - skipping Qdrant for company
ğŸ“ Loading raw text context (3000 chars)
âœ“ Successfully extracted company: World Labs
âœ“ Saved to: data/structured/world-labs.json
â†’ No Qdrant overhead, pure LLM extraction
```

---

## ğŸ”— Integration Points

```
â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”
â”‚                   main() function                         â”‚
â”‚                                                          â”‚
â”‚  1. Parse arguments                                      â”‚
â”‚     â””â”€ Get: --fallback-strategy (or use default)        â”‚
â”‚                                                          â”‚
â”‚  2. Set global FALLBACK_STRATEGY                         â”‚
â”‚     â””â”€ This controls all 6 extraction functions         â”‚
â”‚                                                          â”‚
â”‚  3. Call extraction functions                            â”‚
â”‚     â”œâ”€ extract_company_info()     â† Uses global         â”‚
â”‚     â”œâ”€ extract_events()            â† Uses global        â”‚
â”‚     â”œâ”€ extract_snapshots()         â† Uses global        â”‚
â”‚     â”œâ”€ extract_products()          â† Uses global        â”‚
â”‚     â”œâ”€ extract_leadership()        â† Uses global        â”‚
â”‚     â””â”€ extract_visibility()        â† Uses global        â”‚
â”‚                                                          â”‚
â”‚  4. Each function checks FALLBACK_STRATEGY               â”‚
â”‚     â”œâ”€ 'raw_only'   â†’ Skip Qdrant                       â”‚
â”‚     â”œâ”€ 'qdrant_only'â†’ Fail if no results               â”‚
â”‚     â””â”€ 'qdrant_first'â†’ Fallback to raw                 â”‚
â”‚                                                          â”‚
â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜
```

---

## âœ¨ Summary

**Three Strategies, One Parameter**

| Command | Strategy | Best For | Behavior |
|---------|----------|----------|----------|
| `--fallback-strategy qdrant_first` | Production | â­ Default | Try Qdrant, fallback to raw |
| `--fallback-strategy qdrant_only` | Validation | Strict | Fail if no Qdrant results |
| `--fallback-strategy raw_only` | Debugging | Baseline | Skip Qdrant, use raw text |

**Logging Indicators**
- ğŸ” = Qdrant search initiated
- ğŸ¯ = Results ranked by similarity
- ğŸ“Š = Using Qdrant context
- âš ï¸ = Falling back to raw text
- âš™ï¸ = Strategy-driven action
- âŒ = Error/failure
- âœ… = Success

**All 6 Functions Updated**
- Each enforces the selected strategy
- Each logs appropriately
- Each handles errors for that strategy

---

**Visual Guide v1.0** | **Status: âœ… Complete**
